### PR TITLE
Missed 'open' statements

### PR DESCRIPTION
**Pull request by sam-m888**
From: Dale Athanasias <dalegrind@...>
Mon, 25 Apr 2016 13:41:18 +1000
Subject: Missed 'open' statements

Hi Sam,
Here's a few files with missed 'open' statements:
gramps/gen/filters/_filterlist.py
gramps/plugins/export/exportftree.py
gramps/plugins/database/bsddb_support/write.py

And some older files which you probably left alone for a reason?
windows/nonAIO/builder/build_GrampsWin32.py
windows/nonAIO/check_gtk_install.py
windows/nonAIO/builder/make_launcher.py
windows/nonAIO/builder/check_gtk_install.py
windows/nonAIO/nsis/gcheck.py

Regards

Dale

Re: Prefer with to open files
https://github.com/gramps-project/gramps/pull/113

### CODE REVIEW

**Comment by sam-m888** [comment]
@daleathan Yes I left the `windows/nonAIO/` files alone as they need to be redone, and are kept for reference.

In `gramps/plugins/database/bsddb_support/write.py` left alone the following:

-  line 2540  as no matching close statement
> vers_file = open(bdbversion_file)

- line 2622  as no matching close statement
> backup_table = open(backup_name, 'rb')

**Comment by sam-m888** [comment]
Also found one more `open` statement in:

-  [Line 172](https://github.com/gramps-project/gramps/blob/master/gramps/plugins/importer/importgeneweb.py#L172) for which the mode string looks wrong.
> self.f = open(file, "rUb")

**Comment by eno93** [comment]
For [L2540-L2542](https://github.com/gramps-project/gramps/blob/761cf8f42802cc5056b8afd2ed7587f85143d511/gramps/plugins/database/bsddb_support/write.py#L2540-L2542) change it to:
```
            with open(bdbversion_file) as vers_file:
                bsddb_version = vers_file.readline().strip()
                bsddb_version = ".".join([str(v) for v in safe_eval(bsddb_version)])
```

**Comment by eno93** [comment]
For [L2623-L2624](https://github.com/gramps-project/gramps/blob/761cf8f42802cc5056b8afd2ed7587f85143d511/gramps/plugins/database/bsddb_support/write.py#L2623-L2624) change it to:
```
        with open(backup_name, 'rb') as backup_table:
            load_tbl_txn(database, backup_table, tbl)

```

**Comment by sam-m888** [comment]
@eno93 Thank you, I've added your changes.

Reading the code again I can now see that for the first one the "if" statement checks if the file exist and if true opens it and reads the contents but did not close it. And for the second one it would open and load the backup files but not close them.

**Comment by eno93** [comment]
If you use ack-grep it show the following files also may need to be looked at.

gramps/gen/plug/utils.py
gramps/gen/plug/report/_paper.py
gramps/gen/plug/report/_book.py
gramps/gen/plug/_pluginreg.py
gramps/gen/plug/docbackend/docbackend.py
gramps/gen/db/generic.py
gramps/gen/utils/docgen/csvtab.py
gramps/gen/utils/docgen/odstab.py
gramps/gen/utils/configmanager.py
gramps/test/test_util.py
gramps/cli/grampscli.py
gramps/cli/clidbman.py
gramps/grampsapp.py
gramps/gui/dbman.py
gramps/gui/aboutdialog.py
gramps/gui/widgets/grampletbar.py
gramps/gui/widgets/grampletpane.py
gramps/plugins/docgen/svgdrawdoc.py
gramps/plugins/docgen/rtfdoc.py
gramps/plugins/docgen/odfdoc.py
gramps/plugins/docgen/asciidoc.py
gramps/plugins/tool/verify.py
gramps/plugins/database/test/db_test.py
gramps/plugins/database/bsddb_support/summary.py
gramps/plugins/database/bsddb_support/read.py
gramps/plugins/webreport/webcal.py
gramps/plugins/webreport/narrativeweb.py
gramps/plugins/importer/importprogen.py
gramps/plugins/importer/importgedcom.py
gramps/plugins/importer/importgeneweb.py
gramps/plugins/importer/importxml.py
gramps/plugins/lib/libgedcom.py
gramps/plugins/export/exportcsv.py
gramps/plugins/export/exportftree.py
gramps/plugins/export/exportxml.py
gramps/plugins/export/exportvcalendar.py
gramps/plugins/export/exportgeneweb.py

**Comment by sam-m888** [comment]
@eno93 Ok, thanks for those, I've completed and tested as many as I could for today.

Some notes for myself ( or anyone interested in helping :) )

### Partially done
[1 left]    gramps/plugins/tool/verify.py  **(multiple close statements? With should simplify)**
[1 more?]    gramps/plugins/database/bsddb_support/summary.py  **( see .open( and .close further down page )**
[1 left]    gramps/plugins/importer/importgedcom.py **( further down page)**

### Have def close() methods or similar...
**custom context managers need to check if the close method is used outside of class?**

gramps/gen/plug/docbackend/docbackend.py
gramps/gen/utils/docgen/csvtab.py
gramps/gen/utils/docgen/odstab.py
gramps/plugins/docgen/rtfdoc.py
gramps/plugins/docgen/asciidoc.py
gramps/plugins/webreport/webcal.py
gramps/plugins/docgen/svgdrawdoc.py  **(closed by method "def end_page()" ?)**
gramps/plugins/webreport/narrativeweb.py **(closed by method def close_file() )**

#### Need to examine more closely
gramps/gen/plug/utils.py  **(uses a number of different open statements. looks complicated)**
gramps/cli/clidbman.py **(need to check what's happening?)**
gramps/grampsapp.py  **(ask as seems to redefine builtins depending on OS)**
gramps/plugins/importer/importprogen.py **(not sure)**
gramps/plugins/importer/importgeneweb.py **(no close statement - looks strange eg: "rUb" and encoding further down)**
gramps/plugins/importer/importxml.py  **(uses a " finally:" statement to ensure file closes one of two different open statements)**
gramps/plugins/export/exportxml.py **(two different types of open statement)**
gramps/plugins/export/exportcsv.py **(not sure what's happening as two open statements one reassigned to the other? also see unicodewriter is this from before python3?)**
gramps/test/test_util.py **(Should I touch the test?)**
gramps/plugins/database/test/db_test.py **(Should I touch the test?)**

**Comment by uxcq** [comment]
@sam-m888 What you've done looks good to me.

You might also want to use grep (like suggested above) to also check for:
```
.close(
= zipfile.ZipFile(
= gzip.open(
= gzip.GzipFile(
= tarfile.open(
```
and other statements that could use the "`with`" context manager.

Another thing you may consider looking at is one of my pet peeves with the Gramps source and API that is the use of Java-style setter/getter methods instead of the use of Python [`@property`]( https://docs.python.org/3/library/functions.html#property) decorator.

**Comment by sam-m888** [comment]
@dsblank Doug can you would you be able to please accept what I've done so far? I'll open a new PR for the rest as I do it. uxcq's comment above mentions a separate issue that reminds me of [7335](https://gramps-project.org/bugs/view.php?id=7335) Is it something that should be done/looked at?

**Comment by dsblank** [comment]
Yes, I think it will be good to break these into smaller parts. I see that @eno93 has already looked it over, too. So, I'll look it over and then merge.

File: gramps/plugins/export/exportvcalendar.py [hunk 1]
```diff
@@ -96,7 +96,25 @@ def export_data(self, filename):
...
+                    self.update()
+
+                for key in self.db.iter_family_handles():
+                    self.write_family(key)
+                    self.update()
+
+                self.writeln("")
+                self.writeln("END:VCALENDAR")
+
+                return True
```
Comment (dsblank) [inline gramps/plugins/export/exportvcalendar.py]: Perhaps pull this return out of the with block? We can do that later.

File: gramps/plugins/lib/libgedcom.py [hunk 2]
```diff
@@ -1129,14 +1129,12 @@ def __init__(self):
 
         try:
             filepath = os.path.join(DATA_DIR, "gedcom.xml")
-            ged_file = open(filepath, "rb")
+            with open(filepath, "rb") as ged_file:
+                parser = GedInfoParser(self)
```
Comment (dsblank) [inline gramps/plugins/lib/libgedcom.py]: I'm not sure about having all of this in the try block. And not giving any indication of error if there is a problem. Wonder what that try block is supposed to be doing?

**Comment by dsblank** [comment]
Two small issues that we might want to look at first.

File: gramps/plugins/lib/libgedcom.py (hunk 2 above)
Comment (QuLogic) [inline gramps/plugins/lib/libgedcom.py]: Since it was originally around just the `open`, it's probably checking for non-existent files. Maybe simpler to do `if not os.path.exists(filepath): return`.

File: gramps/plugins/export/exportvcalendar.py (hunk 1 above)
Comment (sam-m888) [inline gramps/plugins/export/exportvcalendar.py]: Ok.

https://github.com/gramps-project/gramps/pull/124/commits/1c0823bc233716cfe1c8975562ffb9a744632f09

File: gramps/plugins/lib/libgedcom.py (hunk 2 above)
Comment (sam-m888) [inline gramps/plugins/lib/libgedcom.py]: @dsblank Should I revert this change or go with @QuLogic 's suggestion?

Comment (dsblank) [inline gramps/plugins/lib/libgedcom.py]: I'd go with @QuLogic and we can test it to make sure that works well. I hate general try/except code with silently handling whatever errors may have happened. Thanks!

Comment (sam-m888) [inline gramps/plugins/lib/libgedcom.py]: Ok

https://github.com/gramps-project/gramps/pull/124/commits/820e981012d5db3457503c2f590db51858c2702e

**Comment by dsblank** [comment]
Thanks for your work!

**Comment by uxcq** [comment]
@sam-m888 Any interest ? In getting rid of [Java-style setter/getter](https://github.com/gramps-project/gramps/pull/124#issuecomment-215306755)?

**Comment by sam-m888** [comment]
@uxcq @dsblank Probably best you as on the Gramps developers mailing list. But my guess is all contributions are welcome :+1:

**Comment by wroldwiedbwe** [comment]
https://github.com/gramps-project/gramps/pull/1068


https://github.com/gramps-project/gramps/pull/113

**Comment by sam-m888** [comment]
more
https://github.com/gramps-project/gramps/pull/1068
